account_invoice_change_currency/tests/test_account_invoice_change_currency.py

```diff
@@ -0,0 +1,320 @@
+                "rate": new_rate,
+                "currency_id": usd.id,
+                "company_id": inv.company_id.id,
+            }
+        )
+        rate = usd_updated_rate.rate
+        inv.write({"custom_rate": rate})
+        inv.action_account_change_currency()
+        expected_value = before_amount
+        # TODO: Check float comparation, 1562.0 vs 1562.0
+        self.assertEqual(
+            inv.amount_total,
+            expected_value,
+            "Total amount of invoice does not equal to expected value!!!",
+        )
+        # change custom rate then we trigger the conversion 2 times
+        # The currency.rate modification above will be ignored and keep the
+        # custom rate
+        old_rate = inv.custom_rate
+        inv.write({"currency_id": usd.id, "custom_rate": rate})
+        inv.action_account_change_currency()
+        before_amount = inv.amount_total
+        rate = usd_updated_rate.rate
+        inv.action_account_change_currency()
+        expected_value = before_amount * rate / old_rate
+        self.assertEqual(
+            float_compare(inv.amount_total, expected_value, self.precision),
+            0,
+            "Total amount of invoice does not equal to expected value!!!",
+        )
+        before_amount = inv.amount_total
+        rate = old_rate + 1
+        inv.write({"custom_rate": rate})
+        inv.action_account_change_currency()
+        expected_value = before_amount * rate / old_rate
+        # TODO: Check float comparation, 4107.93 vs 4107.946074605804
+        self.assertEqual(
+            float_compare(inv.amount_total, expected_value, 1),
+            0,
+            "Total amount of invoice does not equal to expected value!!!",
+        )
+        inv.action_account_change_currency()
+        # TODO: Check float comparation, 4107.93 vs 4107.946074605804
+        self.assertEqual(
+            float_compare(inv.amount_total, expected_value, 1),
+            0,
+            "Total amount of invoice does not equal to expected value!!!",
+        )
+        # Test if there are no currency
+        inv.write({"currency_id": False})
+        self.assertEqual(
+            inv.custom_rate,
+            1.0,
+            "Custom rate of invoice does not equal to expected value!!!",
+        )
+
+    def test_not_currency_change(self):
+        inv = self.create_simple_invoice(inv_type="out_invoice")
+        before_amount = inv.amount_total
+        inv.action_account_change_currency()
+        self.assertEqual(
+            inv.amount_total,
+            before_amount,
+            "Amount must remain the same, because None change was made",
+        )
+
+    def test_old_invoices(self):
+        # This simulate an old invoice (no stored original values)
+        res_currency = self.env["res.currency"]
+        inv = self.create_simple_invoice(inv_type="out_invoice")
+        inv.write({"original_currency_id": False})
+        for line in inv.line_ids:
```

Comment:
Neither loops nor conditionals should be used in tests, as we should know exactly the number of records.

In this case, it's not even necessary because you could write to all lines at once